data/transactions/logic/evalStateful_test.go

```diff
@@ -2795,3 +2874,91 @@ int 77
 `
 	testApp(t, source, ep)
 }
+
+func TestSelfMutateCurrent(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	// start at 9, when such mutation became legal
+	testLogicRange(t, 9, 0, func(t *testing.T, ep *EvalParams, tx *transactions.Transaction, ledger *Ledger) {
+		/* In order to test that apps can now mutate their own app's local state,
+		   we're going to set up a ledger in which an app account is opted into
+		   itself. */
+		ledger.NewLocals(basics.AppIndex(888).Address(), 888)
+		ledger.NewLocal(basics.AppIndex(888).Address(), 888, "hey",
+			basics.TealValue{Type: basics.TealUintType, Uint: 77})
+
+		source := `
+global CurrentApplicationAddress
+byte "hey"
+int 42
+app_local_put
+int 1
+`
+		ed := testApp(t, source, ep)
+		require.Len(t, ed.LocalDeltas, 1)
+		require.Len(t, tx.Accounts, 1) // Sender + 1 tx.Accounts means LocalDelta index should be 2
+
+		require.Contains(t, ed.LocalDeltas, uint64(2))
+		sd := ed.LocalDeltas[2]
+		require.Len(t, sd, 1)
+		require.Contains(t, sd, "hey")
+		require.EqualValues(t, 42, sd["hey"].Uint)
+		require.Len(t, ed.SharedAccts, 1)
+		require.Equal(t, tx.ApplicationID.Address(), ed.SharedAccts[0])
+
+		/* Confirm it worked. */
+		source = `
+global CurrentApplicationAddress
+byte "hey"
+app_local_get
+int 42
+==
+`
+		testApp(t, source, ep)
+
+		source = `
+global CurrentApplicationAddress
+byte "hey"
+app_local_del
+int 1
+`
+		ed = testApp(t, source, ep)
+		require.Len(t, ed.LocalDeltas, 1)
+		require.Len(t, tx.Accounts, 1) // Sender + 1 tx.Accounts means LocalDelta index should be 2
+
+		require.Contains(t, ed.LocalDeltas, uint64(2))
+		sd = ed.LocalDeltas[2]
+		require.Len(t, sd, 1)
+		require.Contains(t, sd, "hey")
+		require.EqualValues(t, basics.DeleteAction, sd["hey"].Action)
+		require.Len(t, ed.SharedAccts, 1)
+		require.Equal(t, tx.ApplicationID.Address(), ed.SharedAccts[0])
+
+		// Now, repeat the "put" test with multiple keys, to ensure only one address is added to SharedAccts
+		source = `
+global CurrentApplicationAddress
+byte "hey"
+int 42
+app_local_put
+global CurrentApplicationAddress
+byte "joe"
+int 21
+app_local_put
+int 1
+`
+		ed = testApp(t, source, ep)
+		require.Len(t, ed.LocalDeltas, 1)
+		require.Len(t, tx.Accounts, 1) // Sender + 1 tx.Accounts means LocalDelta index should be 2
+
+		require.Contains(t, ed.LocalDeltas, uint64(2))
+		sd = ed.LocalDeltas[2]
+		require.Len(t, sd, 2)
+		require.Contains(t, sd, "hey")
+		require.EqualValues(t, 42, sd["hey"].Uint)
+		require.Contains(t, sd, "joe")
+		require.EqualValues(t, 21, sd["joe"].Uint)
```

Comment:
I think I was thinking this would give a briefer, targeted explanation of what exactly went wrong. But looking at them side by side, I like your style better.  I'll check to make show `require.Equal` gives an ok looking diff - I think it will.

Reply:
Yes, it's nice. Will be in the next push.